### PR TITLE
Add simple menu for better UX

### LINKED ISSUE #173: For timezone selections, give options
The current flow is confusing for timezone selection, and the valid options should be provided. US/Eastern isn't exactly an intuitive thing to enter if you don't know what options are available. Maybe do a menu like with keyboard layouts.

### PR DESCRIPTION
**Pull request by svartkanin**
This PR adds the simple menu (https://github.com/IngoMeyer441/simple-term-menu) as a new selection menu to enhance the UX of the guided user flow. 

It has been suggested in issue https://github.com/archlinux/archinstall/issues/655 and this PR changes most selection menus to use the new layout. 

The only part where it still uses the old one is during manual partitioning. The reason for that is that I think there are some bugs in the partitioning code and it kept exploding when I was going through the flow in master. Therefore, I decided to leave that one as is and fix the bugs first and only after switch it over to the new layout as well. 

*Edit: This will fix #173 (I can't link earlier issues to newer PR's because the issues was created before the move from Torxed/archinstall to archlinux/archinstall)*

### CODE REVIEW

**Comment by svartkanin** [comment]
Thanks for the feedback.
All the code in the simple-menu dependency is actually all in a single file. If you want to avoid dependencies then there's the simple option to copy that in manually as well. 
Updates have to be done manually as well then tho.

**Comment by mhussaincov93** [comment]
hi there,
blind user here,
what inpact will this simple menu have on the accessibility of archinstall?
is this simple menu number based?
e.g
1 english, 2 spanish, etc?
best way of testing the accessibility of these changes is to boot from
the arch iso when the boot prompt comes up down arrow once and press
enter it will launch the screen reader.
I hope that accessibility is not broken with this.
Majid

**Comment by svartkanin** [comment]
Hey @Torxed ,
did you manage to test this again? 
Cheers

**Comment by Torxed** [comment]
> Hey @Torxed , did you manage to test this again? Cheers

Not yet, I'll test this one and my other 3-4 PR's I submitted today. Most likely on Wednesday when I'm back at a computer with proper internet. I'm limited to 56k modem speeds for a few days.

**Comment by svartkanin** [comment]
Awesome thanks!

**Comment by Torxed** [comment]
This PR adds a bit of flare to the menu by adding colors and a better interactive selector.
I'm not sure if it's the old if/else we have, but skipping a hard drive will still prompt for wiping or configure individual partitions.

The Text to Speech still works. But it will need some modification to the text output. For instance it's a bit unclear that you can/should use the arrow keys to navigate the menu items, and that you start on the first menu item.
I'm not a frequent TTS user, but this might be an improvement.

screenshot

**Comment by svartkanin** [comment]
Thanks for the review.
I'm happy to include the menu file directly and remove the dependency.
Is there anything else specifically you would like to get changed or updated in this PR?

I'm planning on a next PR to address the remaining old menu selection in the manual partitioning process. 

Also I can play around with displaying already chosen options while going through the menu which could serve as a history in a way. But it would be something in a subsequent PR if there's interest to merge this one and worth the effort that way :)

File: examples/guided.py
Comment (Torxed) [inline examples/guided.py:7]: All API's are exposed via `archinstall.<API call>`, so no need to import `archinstall.lib.hardware` directly, and star imports are generally a bad thing and the linter will hate on me when merging this PR. I'm not sure you're using anything from `hardware.py` either so you can probably omit this line entirely?

File: profiles/desktop.py
Comment (Torxed) [inline profiles/desktop.py:3]: If you can, add `from archinstall.lib.menu import Menu` into [`__init__.py` here](https://github.com/archlinux/archinstall/blob/master/archinstall/__init__.py) instead to expose it in `archinstall.Menu` for convenience when using archinstall in dotfiles.

File: profiles/i3.py
Comment (Torxed) [inline profiles/i3.py:4]: Same here

File: profiles/server.py
Comment (Torxed) [inline profiles/server.py:6]: Same here

**Review by Torxed — CHANGES_REQUESTED** [review]

**Comment by Torxed** [comment]
> Thanks for the review. I'm happy to include the menu file directly and remove the dependency. Is there anything else specifically you would like to get changed or updated in this PR?
> 
> I'm planning on a next PR to address the remaining old menu selection in the manual partitioning process.
> 
> Also I can play around with displaying already chosen options while going through the menu which could serve as a history in a way. But it would be something in a subsequent PR if there's interest to merge this one and worth the effort that way :)

I think that would be good for now, unless someone else has objections *(they might after your finished, it happens)* I don't see why this isn't a better step towards handling menus. Instead of trying to save the previous menu options, perhaps we can have a final menu at the end outputting the result `config.json` data. And tie entries to calling the function responsible for that entry. It would mean keeping and updating a map of what entries comes from which function. But that would allow people to "go back" and re-edit a entry before installing.

Obviously, that would be for the next version of the menu handling perhaps.
For stability reasons, I'll wait until 2.4.0 to include this PR, as I'm already struggling a bit with stability issues re-working a lot of the disk code base. So you would have until the January ISO to do any work on this, if you choose to do so : )

**Comment by svartkanin** [comment]
@Torxed 
I've added the simple menu library as a >= dependency so it will always be updated. To keep it more stable we could either lock it to the current version (==) or copy it into the package. I'm generally leaning towards the locking approach since future updates will be a simple version change, however, the external dependency remains. If we copy it into the package, updating will result in bigger diffs.
I'm happy with either option though, so if the goal is to try to keep the package without dependencies then let's copy it over.

I like the idea of having a final version of the options displayed with the possibility to change them again.
To take it a step further, would it make sense to actually have it as a global menu instead? Meaning, instead of going through the single menus sequentially, we could have a root menu where all these options are displayed and can be edited and after all required ones have been chosen the installation can be triggered?

Just to understand, if you would like to wait until 2.4.0 to include it, would we keep it open until then or is it possible to merge it and just not release it? Not sure how the release process works, if it's taking the most recent master it will be tricky I suppose...

**Comment by Torxed** [comment]
@svartkanin Lets copy it over! : ) Reasoning: Bare in mind that all dependencies has to be a [Arch Linux package](https://archlinux.org/packages/?sort=&q=python-simple&maintainer=&flagged=) for the inclusion to work, we don't use/avoid relying on pip or similar alternatives to include dependencies. There for, the source has to be included in archinstall or be left omitted I believe.
Perhaps there's some automagic to the way requirements gets pulled upon build, but I highly doubt it.
*(keep in mind that this is a Arch Linux package first and foremost, and python library second)*

I'm not sure too which scale that would alter the source code, since the code is written around API calls, and `guided` serves as a decent example of how to build a installer or a templated installation of a system. That would probably mean `guided` stopped being a decent example and became just another monster like any other distro - making it harder to add or contribute to individual steps. This is just my fear, not something to take for granted.

Yes, we would keep it open *(and work on if you wish)* until we'ved released v2.3.0 *(which is bound to happen really soon)*, after which we can happily merged this into master if we deem the PR stable *(which it kinda is with the option we've got here)*. If we merge it into master then any new features or changes to this functionality would be a new PR - but until then it's free for all on this branch and we'll merge once stable again : )

File: profiles/desktop.py
Comment (Torxed) [inline profiles/desktop.py]: Don't forget to change https://github.com/archlinux/archinstall/blob/bdcf33cdc6f6ac4c0bcc8610c064b47a82284eb9/profiles/desktop.py#L46

Into `archinstall.Menu` : )

Comment (Torxed) [reply profiles/desktop.py]: I like the recent changes!

Comment (svartkanin) [reply profiles/desktop.py]: Yes missed that

**Comment by svartkanin** [comment]
@Torxed I've now migrated the partitioning code as well to the new menu and inserted a table for better visibility. Let me know what you think about it :)

**Comment by svartkanin** [comment]
On 
> I think that would be good for now, unless someone else has objections (they might after your finished, it happens) I don't see why this isn't a better step towards handling menus. Instead of trying to save the previous menu options, perhaps we can have a final menu at the end outputting the result config.json data. And tie entries to calling the function responsible for that entry. It would mean keeping and updating a map of what entries comes from which function. But that would allow people to "go back" and re-edit a entry before installing.

Did you mean that after being done with the selection as is now and before the `config.json` will be shown to have an "edit" menu to modify any possible selections? Something like

```
Change lanugate
Change mirrors
...
Continue
```
and then continue with the normal flow?

**Comment by Torxed** [comment]
> On
> 
> > I think that would be good for now, unless someone else has objections (they might after your finished, it happens) I don't see why this isn't a better step towards handling menus. Instead of trying to save the previous menu options, perhaps we can have a final menu at the end outputting the result config.json data. And tie entries to calling the function responsible for that entry. It would mean keeping and updating a map of what entries comes from which function. But that would allow people to "go back" and re-edit a entry before installing.
> 
> Did you mean that after being done with the selection as is now and before the `config.json` will be shown to have an "edit" menu to modify any possible selections? Something like
> 
> ```
> Change lanugate
> Change mirrors
> ...
> Continue
> ```
> 
> and then continue with the normal flow?

Something like that yes.
```python
change_option = archinstall.Menu('Select a step you wish to change or skip and continue with installation', archinstall.arguments.keys(), skip=True).run()
```
And either just delete the key they want to change, and call `ask_user_questions()` once more and it would ask to re-enter that key. Or make all `ask_user_questions()`-questions into function calls individually of each argument, and call that specific function that corresponds to the `change_option` key.

**Comment by dylanmtaylor** [comment]
Can this be rebased on master (or master merged into this PR) so an ISO is generated with the latest changes in it for testing?

**Comment by Torxed** [comment]
> Can this be rebased on master (or master merged into this PR) so an ISO is generated with the latest changes in it for testing?

I can fix in a while, out climbing but I'll give it a go in about two hours if OP doesn't beat me to it :)

**Comment by svartkanin** [comment]
@Torxed thanks for handling the licensing things i wasn't sure if it should be included in the source or not.

I'll try to merge master today (AEST ;) )

**Comment by Torxed** [comment]
> @Torxed thanks for handling the licensing things i wasn't sure if it should be included in the source or not.
> 
> I'll try to merge master today (AEST ;) )

We sure are spread out around the world at this point in this project ^^
Latest master went in without any issues whatsoever in terms of merging.

**Comment by atirut-w** [comment]
Probably too late to suggest but I think we could simply use a library that wraps around the `dialog` command like Pythondialog. It just works and I think it suits the purpose of this installer.

**Comment by dylanmtaylor** [comment]
Needs master merged in again

**Comment by Torxed** [comment]
Done, that one was a bit more tricky. Mainly the latest changes in `user_interaction.py` that I'm a bit worried about.

**Comment by Torxed** [comment]
Fixed and merged in this branch & pr's changes to `ask_for_main_filesystem_format()` with `master` changes properly. Also fixed some imports and removed the last `generic_select()` that I could find.

File: archinstall/lib/simple_menu.py
Comment (dylanmtaylor) [inline archinstall/lib/simple_menu.py:5]: here by -> hereby.

Comment (Torxed) [reply archinstall/lib/simple_menu.py:5]: Thank you, I will correct the phrases!

File: archinstall/lib/menu.py
Comment (dylanmtaylor) [inline archinstall/lib/menu.py:52]: Not a huge fan of this color scheme. Maybe we could go with something close to the Arch blue to match?

File: archinstall/lib/simple_menu.py
Comment (dylanmtaylor) [inline archinstall/lib/simple_menu.py:79]: might need changed here as well.

**Comment by dylanmtaylor** [comment]
@svartkanin Could you merge in https://github.com/svartkanin/archinstall/pull/1? I think the look is a lot better and it makes it match the Arch color scheme.

**Comment by svartkanin** [comment]
@dylanmtaylor That's good, I'll have a look and merge it as soon as I get back to the computer

**Comment by dylanmtaylor** [comment]
I'm wondering how the accessibility of this menu system is with espeak - just saw an issue from a blind user and it made me think of this change

**Comment by Torxed** [comment]
> I'm wondering how the accessibility of this menu system is with espeak - just saw an issue from a blind user and it made me think of this change

I've tried it, so far no issues that I can distinguish.
Works well with espeakup at least.

**Comment by svartkanin** [comment]
@dylanmtaylor I left a comment on the PR

**Comment by dylanmtaylor** [comment]
> @dylanmtaylor I left a comment on the PR

I suppose it makes sense to not touch the upstream file. I made a commit to remove the change to default menu cursor

**Comment by dylanmtaylor** [comment]
> > I'm wondering how the accessibility of this menu system is with espeak - just saw an issue from a blind user and it made me think of this change
> 
> I've tried it, so far no issues that I can distinguish. Works well with espeakup at least.

My biggest concern is how it would handle conveying which item is selected.

**Comment by Torxed** [comment]
> 

Oddly enough, it works near realtime.
At least when I tried it. It aborts the previous TTS when changing item and simply starts mentioning the new selection.
Something I'm not sure even our current method of menu system does. The TTS also remembers "where you are" without having to re-read all the entries which we currently trigger by wiping the screen *(for some of our menus)*.

**Comment by svartkanin** [comment]
@Torxed I saw release 2.3.0 is out, are there any more concerns with this PR or could we consider merging it? 
When it's merged I'd be happy to work on the things we mentioned in the comments, foremost to add a global menu after the configuration to re-edit certain settings.

**Comment by dylanmtaylor** [comment]
> @Torxed I saw release 2.3.0 is out, are there any more concerns with this PR or could we consider merging it? When it's merged I'd be happy to work on the things we mentioned in the comments, foremost to add a global menu after the configuration to re-edit certain settings.

This is currently targeting 2.4.0, I think it'll be the n+2 release, as IIRC a 2.3.1 was planned.

**Comment by Torxed** [comment]
> > @Torxed I saw release 2.3.0 is out, are there any more concerns with this PR or could we consider merging it? When it's merged I'd be happy to work on the things we mentioned in the comments, foremost to add a global menu after the configuration to re-edit certain settings.
> 
> This is currently targeting 2.4.0, I think it'll be part of the n+2 release, as IIRC a 2.3.1 was planned.

Correct. A few things that might need patching. But I could potentially crate a branch of 2.3.0 and cherry-pick certain improvements here and there. I don't think 2.3.1 would be very huge or even relate to v2.4.0. so might be worth doing that, merge this PR into master and all master work goes towards v2.4.0 which will be the larger project?

**Comment by dylanmtaylor** [comment]
> > > @Torxed I saw release 2.3.0 is out, are there any more concerns with this PR or could we consider merging it? When it's merged I'd be happy to work on the things we mentioned in the comments, foremost to add a global menu after the configuration to re-edit certain settings.
> > 
> > 
> > This is currently targeting 2.4.0, I think it'll be part of the n+2 release, as IIRC a 2.3.1 was planned.
> 
> Correct. A few things that might need patching. But I could potentially crate a branch of 2.3.0 and cherry-pick certain improvements here and there. I don't think 2.3.1 would be very huge or even relate to v2.4.0. so might be worth doing that, merge this PR into master and all master work goes towards v2.4.0 which will be the larger project?

I suppose we could make a branch called future-2.3.1 or something and release a version of that which will ship in Jan, and then use master for 2.4.0 development

**Comment by Torxed** [comment]
Merged, I created [v2.3.1-dev](https://github.com/archlinux/archinstall/tree/v2.3.1-dev) for development on the future version v2.3.1 which will contain some minor fixes, and master can continue as usual for "every day development" towards future things to come (v2.4.0 mainly).

**Comment by Torxed** [comment]
@svartkanin Great job on this one, I'm pretty sure it wasn't crystal clear where to change things and replace all the menu options. So thank you for taking the time to dig through the source!

**Comment by svartkanin** [comment]
Sounds good!

**Comment by svartkanin** [comment]
@Torxed it's a really great project and i was very happy to start contributing to it. I'll continue updating things on the menu part but would also like to stqrt picking up other things :)
Is there a project plan or anything the like?

**Comment by Torxed** [comment]
> @Torxed it's a really great project and i was very happy to start contributing to it. I'll continue updating things on the menu part but would also like to stqrt picking up other things :) Is there a project plan or anything the like?

I'm not perfect at updating this, but the milestones should broadly represent things we know we want in specific releases: https://github.com/archlinux/archinstall/milestones

I usually do most of the partitioning work because it's a PITA, but feel free to pick up anything you want :) All PR's are welcome, not all make it into `master` but most of them do.

**Comment by svartkanin** [comment]
Great I'll have a look to see where i can help out

**Pull request merged**
